In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import pandas as pd
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import r2_score
import numpy as np
from sklearn.model_selection import validation_curve
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import cross_val_score

import xgboost as xgb
from sklearn.metrics import mean_squared_error
from xgboost import XGBRegressor


In [16]:
import pandas as pd
from pysal.model import spreg
from pysal.lib import weights

# Load the data
df = pd.read_csv(r"D:\POI data\728_7acc\housing_hasId_reclass3_acc.csv")

# Encode categorical variables
#df_encoded = pd.get_dummies(df, columns=['property_type','tenure'], drop_first=True)
df_encoded = df
# Create spatial weights matrix
coords = df_encoded[['lon', 'lat']].values
w = weights.KNN.from_array(coords, k=5)
w.transform = 'r'  # row-standardization

# Prepare predictors
X = df_encoded.drop(['price', 'address', 
        'Price_per_square_meter', 'id', 'lat', 'lon','borough',
        'property_type','tenure',
             'Eating and drinkingAccess15m',
'Commercial servicesAccess15m',
'EducationAccess15m',
'Manufacturing and productionAccess15m',
     'Bus stopAccess15m',
  'EntertainmentAccess15m',
'Public infrastructureAccess15m',
              'TourismAccess15m'], axis=1)
X = (X - X.mean()) / X.std()  # Standardize the predictors

# Add constant to the predictors
X = pd.DataFrame(np.hstack([np.ones((X.shape[0], 1)), X])).values

# Prepare target variable
y = df_encoded['Price_per_square_meter']
y = (y - y.mean()) / y.std()  # Standardize the target variable

# Fit a spatial error model
ml_error = spreg.ML_Error(y.values[:, None], X, w)

# Print the summary
print(ml_error.summary)


c:\Users\84398\AppData\Local\Programs\Python\Python39\lib\site-packages\libpysal\weights\weights.py:172: UserWarning: The weights matrix is not fully connected: 
 There are 4 disconnected components.
  warnings.warn(message)
c:\Users\84398\AppData\Local\Programs\Python\Python39\lib\site-packages\scipy\optimize\_minimize.py:913: RuntimeWarning: Method 'bounded' does not support relative tolerance in x; defaulting to absolute tolerance.
  warn("Method 'bounded' does not support relative tolerance in x; "


REGRESSION
----------
SUMMARY OF OUTPUT: MAXIMUM LIKELIHOOD SPATIAL ERROR (METHOD = FULL)
-------------------------------------------------------------------
Data set            :     unknown
Weights matrix      :     unknown
Dependent Variable  :     dep_var                Number of Observations:        3737
Mean dependent var  :     -0.0000                Number of Variables   :          16
S.D. dependent var  :      1.0000                Degrees of Freedom    :        3721
Pseudo R-squared    :      0.3121
Sigma-square ML     :       0.353                Log likelihood        :   -3595.367
S.E of regression   :       0.594                Akaike info criterion :    7222.735
                                                 Schwarz criterion     :    7322.352

------------------------------------------------------------------------------------
            Variable     Coefficient       Std.Error     z-Statistic     Probability
-----------------------------------------------------------

---

In [11]:
df = pd.read_csv(r"D:\POI data\728_7acc\housing_hasId_reclass3_acc.csv")

# Perform one-hot encoding on the categorical variables
#df_encoded = pd.get_dummies(df, columns=['property_type'])
df_encoded = df

# Define predictors and target variable
X = df_encoded.drop(['price', 'address', 
        'Price_per_square_meter', 'id', 'borough'], axis=1)
y = df_encoded['Price_per_square_meter']


X

property_type  bedrooms  bathrooms  size_square_meters  \
0             End of Terrace         5          2                 212   
1              Semi-Detached         3          1                 107   
2                   Terraced         3          1                  82   
3                   Terraced         3          1                  73   
4     Semi-Detached Bungalow         2          1                  85   
...                      ...       ...        ...                 ...   
3732               Apartment         2          2                 131   
3733                    Flat         3          2                 190   
3734                    Flat         1          1                  52   
3735                    Mews         3          2                 106   
3736               Apartment         1          1                  46   

            lat       lon     tenure  garden  AccommodationAccess15m  \
0     51.540268  0.079324   Freehold       1                       0   
1     51.539742  0.127662   Freehold       1                       0   
2     51.542795  0.172597   Freehold       0                       0   
3     51.556139  0.148446   Freehold       1                       1   
4     51.576866  0.129566   Freehold       1                       0   
...         ...       ...        ...     ...                     ...   
3732  51.500537 -0.169462  Leasehold       1                      10   
3733  51.512159 -0.172806  Leasehold       0                     168   
3734  51.514094 -0.180396  Leasehold       1                     123   
3735  51.514029 -0.178216   Freehold       0                     130   
3736  51.522983 -0.183308  Leasehold       1                     108   

      AirportsAccess15m  ...  Bus stopAccess15m  CultureAccess15m  \
0                     0  ...                 20                 0   
1                     0  ...                 55                 1   
2                     0  ...                 12                 0   
3                     0  ...                 33                 0   
4                     0  ...                  7                 0   
...                 ...  ...                ...               ...   
3732                  0  ...                 42                16   
3733                  0  ...                126                52   
3734                  0  ...                 43                16   
3735                  0  ...                 77                23   
3736                  0  ...                150                17   

      EntertainmentAccess15m  GamblingAccess15m  Green parkAccess15m  \
0                          3                  6                    1   
1                          4                  9                    4   
2                          0                  2                    2   
3                          0                  1                    1   
4                          0                  1                    2   
...                      ...                ...                  ...   
3732                       5                  3                    4   
3733                      63                 24                    4   
3734                       2                  6                    1   
3735                      16                 13                    3   
3736                      36                 20                   26   

      Public infrastructureAccess15m  Railway stationsAccess15m  \
0                                 30                          3   
1                                 91                          1   
2                                 20                          0   
3                                 53                          0   
4                                 36                          0   
...                              ...                        ...   
3732                              80                          0   
3733                             366       

In [12]:
from pysal.lib import weights
from pysal.model import spreg

# Combine 'lat' and 'lon' into a 2D numpy array
coords = X[['lon', 'lat']].values
# Create a KNN spatial weight
w = weights.KNN.from_array(coords, k=5)
w.transform = 'r'  # row-standardization

# Convert X and y to numpy arrays
X_np = X.values
y_np = y.values.reshape((-1, 1))

# Run the spatial error model
model = spreg.GM_Error(y_np, X_np, w=w, name_x=X.columns.tolist(), name_y='Price_per_square_meter')

# Print the summary
print(model.summary)


c:\Users\84398\AppData\Local\Programs\Python\Python39\lib\site-packages\libpysal\weights\weights.py:172: UserWarning: The weights matrix is not fully connected: 
 There are 4 disconnected components.
  warnings.warn(message)


TypeError: can only concatenate str (not "int") to str

In [ ]:
data = {
    'Variable': ['CONSTANT', 'bedrooms', 'bathrooms', 'size_square_meters', 'lat', 'lon', 'AccommodationAccess15m', 'AirportsAccess15m', 'Culture and tourismAccess15m', 'Eating and drinkingAccess15m', 'EducationAccess15m', 'EntertainmentAccess15m', 'HealthAccess15m', 'Infrastructure and facilitiesAccess15m', 'Manufacturing and productionAccess15m', 'OfficeAccess15m', 'Public transportAccess15m', 'Road and railAccess15m', 'ServiceAccess15m', 'ShoppingAccess15m', 'Water transportAccess15m'],
    'Coefficient': [44824.6612676, -378.2754568, 769.1812736, -0.3345708, -685.1517479, -502.0382665, -5.7215771, 185.6969145, 15.4173706, -1.0109978, 17.5090073, -29.5113216, 1.0963004, -18.7088178, -18.2949289, 42.1420862, 6.5737330, -4.2325132, -1.1014293, 9.9917338, 4.1988936],
    'Std.Error': [34325.9555096, 58.8144243, 71.7559080, 0.7844160, 666.6632406, 681.6034048, 10.1843527, 364.8417897, 6.1799001, 1.9065770, 11.2410312, 10.6656981, 3.9670999, 5.4739910, 6.5128915, 9.5069267, 5.6670016, 13.0765172, 0.9608790, 1.8474973, 19.7420190],
    'z-Statistic': [1.3058533, -6.4316783, 10.7194138, -0.4265221, -1.0277329, -0.7365548, -0.5618008, 0.5089793, 2.4947605, -0.5302685, 1.5575980, -2.7669376, 0.2763481, -3.4177655, -2.8090333, 4.4327770, 1.1600020, -0.3236728, -1.1462726, 5.4082536, 0.2126882],
    'Probability': [0.1916025, 0.0000000, 0.0000000, 0.6697274, 0.3040755, 0.4613931, 0.5742518, 0.6107667, 0.0126042, 0.5959258, 0.1193286, 0.0056586, 0.7822808, 0.0006314, 0.0049691, 0.0000093, 0.2460480, 0.7461858, 0.2516824, 0.0000001, 0.8315702]
}
pd.set_option('display.float_format', '{:.8f}'.format)
summary_df = pd.DataFrame(data)
summary_df

Variable    Coefficient      Std.Error  \
0                                 CONSTANT 44824.66126760 34325.95550960   
1                                 bedrooms  -378.27545680    58.81442430   
2                                bathrooms   769.18127360    71.75590800   
3                       size_square_meters    -0.33457080     0.78441600   
4                                      lat  -685.15174790   666.66324060   
5                                      lon  -502.03826650   681.60340480   
6                   AccommodationAccess15m    -5.72157710    10.18435270   
7                        AirportsAccess15m   185.69691450   364.84178970   
8             Culture and tourismAccess15m    15.41737060     6.17990010   
9             Eating and drinkingAccess15m    -1.01099780     1.90657700   
10                      EducationAccess15m    17.50900730    11.24103120   
11                  EntertainmentAccess15m   -29.51132160    10.66569810   
12                         HealthAccess15m     1.09630040     3.96709990   
13  Infrastructure and facilitiesAccess15m   -18.70881780     5.47399100   
14   Manufacturing and productionAccess15m   -18.29492890     6.51289150   
15                         OfficeAccess15m    42.14208620     9.50692670   
16               Public transportAccess15m     6.57373300     5.66700160   
17                  Road and railAccess15m    -4.23251320    13.07651720   
18                        ServiceAccess15m    -1.10142930     0.96087900   
19                       ShoppingAccess15m     9.99173380     1.84749730   
20                Water transportAccess15m     4.19889360    19.74201900   

    z-Statistic  Probability  
0    1.30585330   0.19160250  
1   -6.43167830   0.00000000  
2   10.71941380   0.00000000  
3   -0.42652210   0.66972740  
4   -1.02773290   0.30407550  
5   -0.73655480   0.46139310  
6   -0.56180080   0.57425180  
7    0.50897930   0.61076670  
8    2.49476050   0.01260420  
9   -0.53026850   0.59592580  
10   1.55759800   0.11932860  
11  -2.76693760   0.00565860  
12   0.27634810   0.78228080  
13  -3.41776550   0.00063140  
14  -2.80903330   0.00496910  
15   4.43277700   0.00000930  
16   1.16000200   0.24604800  
17  -0.32367280   0.74618580  
18  -1.14627260   0.25168240  
19   5.40825360   0.00000010  
20   0.21268820   0.83157020

In [ ]:
summary_df_Probability = summary_df[summary_df['Probability'] < 0.05]
summary_df_Probability

Variable   Coefficient   Std.Error  \
1                                 bedrooms -378.27545680 58.81442430   
2                                bathrooms  769.18127360 71.75590800   
8             Culture and tourismAccess15m   15.41737060  6.17990010   
11                  EntertainmentAccess15m  -29.51132160 10.66569810   
13  Infrastructure and facilitiesAccess15m  -18.70881780  5.47399100   
14   Manufacturing and productionAccess15m  -18.29492890  6.51289150   
15                         OfficeAccess15m   42.14208620  9.50692670   
19                       ShoppingAccess15m    9.99173380  1.84749730   

    z-Statistic  Probability  
1   -6.43167830   0.00000000  
2   10.71941380   0.00000000  
8    2.49476050   0.01260420  
11  -2.76693760   0.00565860  
13  -3.41776550   0.00063140  
14  -2.80903330   0.00496910  
15   4.43277700   0.00000930  
19   5.40825360   0.00000010

"Culture and tourismAccess15m" 的系数是 15.41737060，表示每增加一个单位的"Culture and tourismAccess15m"，预期的"Price_per_square_meter"将增加大约15.41737060个单位。这里，"单位"的含义取决于"Culture and tourismAccess15m"具体表示的是什么。通常，这可能是某种度量的数量，例如，如果"Culture and tourismAccess15m"表示的是某个房屋周围15分钟步行范围内的文化和旅游设施的数量，那么这个系数就表示每增加一个这样的设施，预期的每平方米房价将增加大约15.41737060。

由于p值<0.05，说明"Culture and tourismAccess15m"的影响是显著的，即我们有足够的证据支持"Culture and tourismAccess15m"对每平方米的房价有影响。